### ISSUE TITLE
Support for `as` prop

### ISSUE THREAD

**Issue by nahtnam** [fgnass/classname-variants#3]
It would be useful to have support for `as` props. For example, for this component:

```
export const SquareButton = styled('button', {
  base: '',
  variants: {
    size: {
      xs: '',
    },
  },
});
```

It would be cool if we could do `<SquareButton as="a" size="xs" />` similar to StyledComponents

**Comment by kndwin** [comment]
Here's my current workaround (I added a `tw` prop as well cause it's clearer that these classes are tailwind + it's shorter)

```typescript
import * as React from "react";
import { variantProps, VariantPropsOf } from "classname-variants/react";
import clsx from "clsx";

const textVariantProps = variantProps({
  base: "text-black",
  variants: {
    b: { true: "font-bold" },
  },
});

interface TextProps<T extends React.ElementType> {
  as?: T;
  tw: string;
  children?: React.ReactNode;
}

export function Text<T extends React.ElementType = "p">({
  as,
  tw,
  ...props
}: TextProps<T> &
  VariantPropsOf<typeof textVariantProps> &
  Omit<React.ComponentPropsWithoutRef<T>, keyof TextProps<T>>) {
  const Component = as ?? "p";
  return (
    <Component
      {...textVariantProps(props)}
      className={clsx(tw, textVariantProps(props).className)}
    />
  );
}
```

Usage
```typescript
export default function HomePage () {
  return (
    <Text as="h1" b tw="text-2l">Hello world</Text>
  )
}
```

**Comment by fgnass** [comment]
I added support for an `as` prop in the `polymorphic-components` branch: https://github.com/fgnass/classname-variants/compare/main...polymorphic-components

It uses the same trick as https://github.com/kripod/react-polymorphic-box and works as expected. The only downside is that the types displayed in the IDE now look rather ugly:

<img width="753" alt="Screenshot 2022-10-20 at 01 27 31" src="https://user-images.githubusercontent.com/89450/196823746-7ad5a100-59ac-4a2a-a911-249ef9ede556.png">

Before this change it used to look like this:

<img width="678" alt="Screenshot 2022-10-20 at 01 27 50" src="https://user-images.githubusercontent.com/89450/196823856-6451cd61-eb19-40eb-8dfc-196fe95f0980.png">

What do you think, is this acceptable?

**Comment by Code-Victor** [comment]
I guess that is the cost of getting polymorphism. it's acceptable 💯
Good work👍

**Comment by kevin-nguyen-mirvac** [comment]
🚀 This is awesome!, would reduce my custom ui library code by 50% T_T

Just a nitpick: notice a potential typo 
https://github.com/fgnass/classname-variants/compare/main...polymorphic-components#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R157

**Comment by fgnass** [comment]
Landed in classname-variants@1.3.0
